Address SR-14920 and add FileDescriptorPort on Linux platform

smumriak opened the pull request:
Proposed change fixes an issue described in SR-14920. Also this pull request adds implementation of `CFFileDescriptor` on Linux platform to match the `CFFileDescriptor` from Darwin and it also adds the Swift counterpart type `FileDescriptorPort` that allows usage of `CFFileDescriptor` without directly linking to CoreFoundation.
While fixing SR-14920 I've also added a typedef called `port_t` that encapsulates the type of fd per platform. On Darwin it's `mach_port_t`, in Linux it's `int`, on Windows it's `HANDLE` and on every other platform it's `void *`. Before this change on Linux all the `__CFPort` variables in `CFRunLoop` were `void *` which would create a logical issue when `CFRunLoop` code would implicitly downcast `void *` to `int` when setting up epoll file descriptors. Also on Windows it's semantically correct to use `HANDLE` type in this case, but I have no windows machine to test this. Maybe @compnerd could help with that.
I did my best to implement `CFFileDescriptor` according to documentation for `CFFileDescriptor` on Darwin, tho I've added additional convenience function `CFFileDescriptorSetFlags`.
`FileDescriptorPort` is intended to bring same value to `CFFileDescriptor` as `NSMachPort` does for `CFPort`, i.e. allow usage of file descriptor input/output events from swift (in this case, I don't think we care about objective-c on Linux). I've reused the existing CF runtime metadata for `CFFileDescriptor`.
One additional point to make is that epoll itself has a bug in kernel which makes all nested epoll file descriptors in edge-trigged mode behave like level-triggered. the fix for this issue is not landed, details can be found in https://lkml.org/lkml/2019/9/2/17. Because of that the code adds the fd to be listened every time client enables callbacks and removes it every time perform callback is called from `CFRunLoop`. This way the code can mimic behavior of `CFFileDescriptor` from Darwin.
In order to make unit tests for `FileDescriptorPort` I made a workaround to load epoll functions via `dlsym` because I could not find a simple way to pass epoll headers to Foundation tests targets, so it would be visible by compiler. I would appreciate any help in this matter.
Unfortunately, I can not run full test suite on current main, because it fails to compile on my machine (compiler can not generate Codable/Decodable for some types in AttributedString implementation). I did ran tests while rebased on `swift-DEVELOPMENT-SNAPSHOT-2021-08-14-a` tho.
Last thing: in order to test the fix for SR-14920 I've used the same sample project I linked in the bug, but preloaded the Foundation built with my changes like so:
```
LD_PRELOAD=/path/to/swift-project/build/Ninja-ReleaseAssert/foundation-linux-x86_64/lib/libFoundation.so swift run
```

smumriak commented:
@swift-ci please test

mickeyl commented:
Great stuff, going to give that a try. Whom do we need to pester to review and, eventually, pull that?

tomerd commented:
cc @compnerd @weissi @Lukasa @drexin

millenomi commented:
@swift-ci please test

millenomi commented:
Real, non-rhetorical question: what projects are relying on `{CF,}RunLoop` on the downstream?

millenomi commented:
> Unfortunately, I can not run full test suite on current main, because it fails to compile on my machine (compiler can not generate Codable/Decodable for some types in AttributedString implementation). I did ran tests while rebased on swift-DEVELOPMENT-SNAPSHOT-2021-08-14-a tho.

Please file a SR bug for this? Also, cc @iCharlesHu.

mickeyl commented:
> Real, non-rhetorical question: what projects are relying on `{CF,}RunLoop` on the downstream?

This question comes always when we talk about using `RunLoop` on Linux programs. Let me tell you my story here: 

I'm working on [automotive diagnosis apps and frameworks using OBD2 adapters](https://github.com/Automotive-Swift). Communication with those adapters works a bit like with AT modems, it's a serial communication stream encapsulated in a transport technology.

There are various adapters out there, some TTY (IOStream), some WiFi (TCP stream), some Bluetooth 3.x (EA + Input/OutputStream), some Bluetooth 4.x (one or two CBCharacteristics with my CBService -> IOStream bridge).

The common denominator is always a pair of Input/Output stream, hence I based my adapter abstraction on Streams. Now the ExternalAccessory and CoreBluetooth stuff clearly is Apple-only, but my TTY, BLE 3.x, and WiFi streams should run on non-Apple platforms as well. The natural way to deal with these streams seems to be a RunLoop and that works great on Apple platforms, but not so much on Linux – yet.

smumriak commented:
> Please file a SR bug for this? Also, cc @iCharlesHu.

sure, I'll file a bug

> Real, non-rhetorical question: what projects are relying on {CF,}RunLoop on the downstream?

I remember there was a swift forums thread started by you asking the very same question. I'm using `RunLoop` in my UI framework as a main loop because it's simple, reliable and easy to extend. But more importantly, it allows me to have my sources to be processed before main queue gets processed on main thread. This way I can easily integrate runloop from GTK, for example, by adding it's fd as a source into RunLoop and call to `gtk_main_iteration` in the process callback. This way the API becomes easy to extend by injecting whatever events source that is not cowered by swift's Foundation

millenomi commented:
Understood. Happy to take this on a few conditions:
- Tests passing, of course.
- The public API should remain as-is, as much as possible? It's out of scope and bad for future evolution to allow downstream clients to rely on things that do not exist on Darwin, for a whole host of reasons. CFFileDescriptor is already API on Darwin, but the Foundation additions are not — and if I merged your header change, ABI would be compromised on Darwin for run loop sources.

smumriak commented:
> Tests passing, of course.
not even questioned :)

> if I merged your header change, ABI would be compromised on Darwin for run loop sources.

you are talking about the `CFRunLoop.h` change, right? Not the `CFFileDescriptor`, since both header and implementation are hidden behind `#if TARGET_OS_LINUX`

smumriak commented:
Swift test pipeline finished with segfault in those tests that I've added. I assume it's because runtime can not load the functions pointers `eventfd`, `eventfd_read` and `eventfd_write`. Swift does not expose them in Glibc module, so I've made a workaround with `dlsym`. Chances are it does not as I expected on swift CI servers. Any ideas how to get those functions in tests more elegantly? In the meantime, I'll try to reset my environment and rerun tests

smumriak commented:
scratch that. i've rebased on latest develop delease tag and was able to build and debug tests. the cause of failed test is missing `CFRetain` after some code cleanup. I still would like to address the header change. Am I reading your comment correctly that I should leave the implicit cast from `void *` to `int` in place?

smumriak commented:
@swift-ci please test

millenomi commented:
@swift-ci please test

smumriak commented:
Looks like a test in docc failed. I'm curious if this change caused considering how it's at the lowest level of foundation. Is there a way to run the complete set of tests locally like CI does? Looks like `./utils/run-test` utility from swift repo runs only tests of stdlib

smumriak commented:
Lol, nevermind, the failing test was disabled 14 minutes ago: https://github.com/apple/swift-docc/commit/162c873bd9cf05182a71164c4ed8514e03f471b8

millenomi commented:
@swift-ci please test

millenomi commented:
@smumriak I need to go through it real quick, but how Linux-specific is the CFFileDescriptor design?

millenomi commented:
Nevermind, I see epoll usage.

smumriak commented:
@millenomi CFFileDescriptor implementation in this PR is very linux specific. FileDescriptorPort swift class tho can be easily adapted to be Darwin compatible with CFFileDescriptor from there by literally removing that convenience function I've added in CFFileDescriptor here

smumriak commented:
I guess what I'm trying to say here is "if you are going to merge this PR could you please wait a bit and lets discuss how I could make it as safe as possible". i.e. i'm asking for your feedback on `CFFileDescriptorSetFlags` and that thing you've mentioned earlier about the header file changes

smumriak commented:
Tiny update: 
- used `_cfStringUniquingKnown` instead of `.rawValue._cfObject`, otherwise `CFRunLoop` would not correctly add port into portset
- removed notion of `port_t,` so public interface does not change on Darwin, but its still using native port alternatives on Linux and windows (`int` and `HANDLE` respectively). this should not break ABI compatibility on windows, but can break it on Linux, tho I think we don't care about that since run loop sources on Linux did not work correctly
- changed run loop mode to `.common` in tests
- removed convenience APIs from `CFFileDescriptor` making it source-compatible with original on Darwin. I'll leave decision to expose `FileDescriptorPort` swift class on Darwin platforms up to you
- docc tests on Linux are still failing and I can not build docc locally (I'm trying to verify that it's not caused by my environment). building toolchain without docc works, tho I'd like CI to run tests on change again

mickeyl commented:
Now that the [merge window for 5.5.3 is open](https://forums.swift.org/t/development-open-for-swift-5-5-3-for-linux-and-windows/54553), is there any chance to get this change in?

tomerd commented:
> Now that the merge window for 5.5.3 is open, is there any chance to get this change in?

the process to get this into 5.5.x would be to first get this approved and merged in the main branch, then prepare a cherry-pick into the release/5.5 branch

drexin commented:
> > Now that the merge window for 5.5.3 is open, is there any chance to get this change in?
> 
> the process to get this into 5.5.x would be to first get this approved and merged in the main branch, then prepare a cherry-pick into the release/5.5 branch

And `release/5.6` as well.

smumriak commented:
Makes sense. Let me rebase on latest master and run tests locally cause it's been more than a month since I did it last time :)

smumriak commented:
updated to latest available main branch. swift-corelibs-foundation tests passing locally

mickeyl commented:
Excellent. Is there anything beyond a review holding this this back to land in `main`?

smumriak commented on Sources/Foundation/FileDescriptorPort.swift:
```diff
@@ -0,0 +1,225 @@
+
+    let kCFRunLoopSourceOrder: CFIndex = 600
+
+    public protocol FileDescriptorPortDelegate: PortDelegate {
```
The goal of this swift class is to give clients of the API ability to use file descriptors with RunLoop class, since CFRunLoop as API is going away (even tho it's still available on Linux)
From your comment I assume it's not possible to enable this in Darwin due to reasons other than source compatibility, right? Asking because on source code and functional levels CFFileDescriptor on Darwin and this implementation are pretty much the same

smumriak replied:
i just realized: this repository is never merged into apple private code. am i right?

mickeyl replied:
Can you perhaps ship this as an external package? Swift-corelibs-foundation seems to be more guarded than linux mainline, it‘s becoming ridiculous.

smumriak commented:
Alright, I've removed FileDescriptorPort swift class completely. This PR now only supplies the fix for SR-14920 and implementation of CFFileDescriptor. There are also tests for CFFileDescriptor, tho since CoreFoundation has no published tests I've added those in place of previously present tests for FileDescriptorPort.

@mickeyl Answering your question if FileDescriptorPort can be published somewhere else. Yes, but making it crossplatform will be impossible without access to CoreFoundation.

smumriak commented:
rebased on latest main, fixed merge conflict and ran tests. gentle ping @millenomi, any chance we could make a decision on this pull request? maybe @parkera or @etcwilde could help with review?

smumriak commented:
This PR is not going to be merged. According to [this post](https://forums.swift.org/t/what-s-next-for-foundation/61939) foundation is no longer going to be using RunLoop as an entity, so anything that makes integration with it or fixes bugs in RunLoop is not going to be needed. I'm closing this pull request
